mlir/lib/Catalyst/Transforms/DetensorizeSCFPass.cpp

```diff
@@ -0,0 +1,163 @@
+            rewriter.replaceUsesWithIf(
+                forOp->getResult(index), from_elem_op->getResult(0),
+                [&](OpOperand &op) { return !isa<tensor::FromElementsOp>(op.getOwner()); });
+        }
+
+        rewriter.replaceOp(forOp, newForOp);
+    }
+
+    void detensorizeIfOp(scf::IfOp if_op, IRRewriter &rewriter)
+    {
+        auto isScalarTensor = [](Value v) {
+            if (!::llvm::isa<TensorType>(v.getType())) {
+                return false;
+            }
+            return dyn_cast<TensorType>(v.getType()).getRank() == 0;
+        };
+
+        if_op->walk([&](scf::YieldOp yield_op) {
+            // Loop over results
+            std::size_t i_result = 0;
+            for (mlir::Value operand : yield_op.getOperands()) {
+                if (isScalarTensor(operand)) {
+                    auto if_result = if_op->getResult(i_result);
+                    // Detensorize operand: extract tensor element before yielding
+                    {
+                        rewriter.setInsertionPoint(yield_op);
+                        mlir::Operation *extract_op = rewriter.create<tensor::ExtractOp>(
+                            yield_op->getLoc(), operand, ValueRange{});
+                        rewriter.replaceUsesWithIf(
+                            operand, extract_op->getResult(0),
+                            [&](OpOperand &op) { return op.getOwner() == yield_op; });
+                        const mlir::Type type = extract_op->getResult(0).getType();
+                        rewriter.modifyOpInPlace(
+                            yield_op, [&]() { yield_op->getOperand(i_result).setType(type); });
+                        rewriter.modifyOpInPlace(if_op, [&]() { if_result.setType(type); });
+                    }
+                    // Retensorize operand: reconstruct tensor after the for body
+                    {
+                        rewriter.setInsertionPointAfter(if_op);
+                        auto from_elem_op = rewriter.create<tensor::FromElementsOp>(
+                            if_op->getLoc(), RankedTensorType::get({}, if_result.getType()),
+                            if_result);
+                        rewriter.replaceUsesWithIf(
+                            if_result, from_elem_op->getResult(0), [&](OpOperand &op) {
+                                return !isa<tensor::FromElementsOp>(op.getOwner());
+                            });
+                    }
+                }
+                i_result += 1;
+            }
+        });
+    }
+
+    void runOnOperation() override
+    {
+        mlir::Operation *root = getOperation();
+        IRRewriter rewriter(root->getContext());
+
+        // llvm::errs() << *root << '\n';
+
+        root->walk([&](scf::ForOp for_op) { detensorizeForOp(for_op, rewriter); });
+
+        // llvm::errs() << *root << '\n';
+
+        root->walk([&](scf::IfOp if_op) { detensorizeIfOp(if_op, rewriter); });
```

Comment:
This is one option, albeit quite a "manual" one 😅  Erick's suggestion of providing rewrite patterns would be the typical approach to go about this I think. In the pass itself you would then invoke a pattern applicator (like `applyPatternsAndFoldGreedily`.
On the other hand, performing a single walk if we know a pattern only needs to be applied once could be more efficient, I'm not sure 🤔 

Regarding your concern about having to modify IR around the if/for op, to me it doesn't look like this would be a problem. Replacing a matched operation with a set of other operations (that is inserting new ops into the parent) is exactly what patterns are for, and that necessarily includes replacing Value uses of the old op with new Values (affecting operands of other ops). 

MLIR lists these [restrictions](https://mlir.llvm.org/docs/PatternRewriter/#restrictions) on patterns:
> Within the match section of a pattern, the following constraints apply:
>  - No mutation of the IR is allowed.
>
> Within the rewrite section of a pattern, the following constraints apply:
> - All IR mutations, including creation, must be performed by the given PatternRewriter. This class provides hooks for performing all of the possible mutations that may take place within a pattern. For example, this means that an operation should not be erased via its erase method. To erase an operation, the appropriate PatternRewriter hook (in this case eraseOp) should be used instead.
> - The root operation is required to either be: updated in-place, replaced, or erased.

But the following [restrictions](https://mlir.llvm.org/docs/PassManagement/#operation-pass) apply on passes:
> - Must not inspect the state of operations that are siblings of the current operation. Must neither access operations nested under those siblings.
>   - Other threads may be modifying these operations in parallel.
>   - Inspecting the state of ancestor/parent operations is permitted.
> - Must not modify the state of operations other than the operations that are nested under the current operation. This includes adding, modifying or removing other operations from an ancestor/parent block.
>   - Other threads may be operating on these operations simultaneously.
>   - As an exception, the attributes of the current operation may be modified freely. This is the only way that the current operation may be modified. (I.e., modifying operands, etc. is not allowed.)

Here it's important that only the operation on which the pass is applied is modifed.